internal/users/localentries/lockedentries.go

```diff
@@ -0,0 +1,369 @@
+	}
+
+	l.localUserEntries, err = parseLocalPasswdFile(l.options.inputPasswdPath)
+	return l.localUserEntries, err
+}
+
+// GetLocalGroupEntries gets the local group entries.
+func (l *WithLock) GetLocalGroupEntries() (entries []types.GroupEntry, err error) {
+	l.mu.Lock()
+	defer l.mu.Unlock()
+
+	l.mustBeLocked()
+
+	if l.localGroupEntries != nil {
+		return l.localGroupEntries, nil
+	}
+
+	l.localGroupEntries, l.localGroupInvalidEntries, err = parseLocalGroups(
+		l.options.inputGroupPath)
+	return l.localGroupEntries, err
+}
+
+// updateLocalGroupEntriesCache updates the local group entries.
+func (l *WithLock) updateLocalGroupEntriesCache(entries []types.GroupEntry) {
+	l.mu.Lock()
+	defer l.mu.Unlock()
+
+	l.mustBeLocked()
+
+	l.localGroupEntries = types.DeepCopyGroupEntries(entries)
+}
+
+// lockGroupFile locks the read/write operation on the group file and returns
+// an unlock function.
+func (l *WithLock) lockGroupFile() (unlock func()) {
+	l.localGroupsMu.Lock()
+	return l.localGroupsMu.Unlock
+}
+
+// IsUniqueUserName returns if an user exists for the given name.
+func (l *WithLock) IsUniqueUserName(name string) (unique bool, err error) {
+	l.MustBeLocked()
+
+	users, err := l.GetUserEntries()
+	if err != nil {
+		return false, err
+	}
+
+	// Let's try to check first the (potentially) cached entries.
+	if slices.ContainsFunc(users, func(u types.UserEntry) bool {
+		return u.Name == name
+	}) {
+		return false, nil
+	}
+
+	// In case we found no user, we need to still double check NSS by name.
+	_, err = user.Lookup(name)
+	if err == nil {
+		return false, nil
+	}
+
+	var userErr user.UnknownUserError
+	if !errors.As(err, &userErr) {
+		return false, err
+	}
+
+	return true, nil
+}
+
+// IsUniqueGroupName returns if an user exists for the given UID.
+func (l *WithLock) IsUniqueGroupName(group string) (unique bool, err error) {
+	l.MustBeLocked()
+
+	// Let's try to check first the (potentially) cached entries.
+	groups, err := l.GetGroupEntries()
+	if err != nil {
+		return false, err
+	}
+
+	if slices.ContainsFunc(groups, func(g types.GroupEntry) bool {
+		return g.Name == group
+	}) {
+		return false, nil
+	}
+
+	// In case we found no user, we need to still double check NSS by name.
+	_, err = user.LookupGroup(group)
+	if err == nil {
+		return false, nil
+	}
+
+	var groupErr user.UnknownGroupError
+	if !errors.As(err, &groupErr) {
+		return false, err
+	}
+
+	return true, nil
+}
+
+// IsUniqueUID returns if an user exists for the given UID.
+func (l *WithLock) IsUniqueUID(uid uint32) (unique bool, err error) {
+	l.MustBeLocked()
+
+	users, err := l.GetUserEntries()
+	if err != nil {
+		return false, err
+	}
+
+	// Let's try to check first the (potentially) cached entries.
+	if slices.ContainsFunc(users, func(u types.UserEntry) bool {
+		return u.UID == uid
+	}) {
+		return false, nil
+	}
+
+	// In case we found no user, we need to still double check NSS by ID.
+	_, err = user.LookupId(strconv.FormatUint(uint64(uid), 10))
```

Comment:
Looking at the implementation of `user.LookupId` (if cgo is enabled), I wonder if it's buggy, because it does not treat all errors which can be returned by `getpwuid` in the "user not found" case as such. According to the man page:

> ERRORS
>        0 or ENOENT or ESRCH or EBADF or EPERM or ...
>               The given name or uid was not found.

The code only treats 0 (nil) and `ENOENT` as "user not found": https://github.com/golang/go/blob/go1.24.4/src/os/user/cgo_lookup_unix.go#L61-L63

Reply:
well, at least we're returning other errors, so we won't end up assigning IDs which are already in use

Reply:
Yeah, I see... Well adding that check is easy, let me know if you want it in

Reply:
no, I mean for how we're acting on the result, the implementation of `user.LookupId` might actually be safer. so let's keep it the way it is and see if we get any unexpected errors which should be treated as "user not found"